### ISSUE TITLE
Support for AgglomerativeClustering

### ISSUE THREAD

**Issue by kota7** [jpmml/sklearn2pmml#396]
Will it be possible to support AgglomerativeClustering?
Below is my try, which says the algorithm is not supported.

```python
import numpy as np
import pandas as pd
from sklearn.cluster import AgglomerativeClustering
from sklearn2pmml import PMMLPipeline, sklearn2pmml

X = np.array([[1, 2], [1, 4], [1, 0],
              [4, 2], [4, 4], [4, 0]])
X = pd.DataFrame(X)
X.columns = ["x1", "x2"]
pipe = PMMLPipeline([("clustering", AgglomerativeClustering())])
pipe.fit(X)

sklearn2pmml(pipe, "clustering.pmml")
```

```shell
Standard output is empty
Standard error:
Exception in thread "main" java.lang.IllegalArgumentException: The transformer object (Python class sklearn.cluster._agglomerative.AgglomerativeClustering) is not a supported Transformer
	at org.jpmml.python.CastFunction.apply(CastFunction.java:47)
	at sklearn.pipeline.Pipeline$1.apply(Pipeline.java:109)
	at sklearn.pipeline.Pipeline$1.apply(Pipeline.java:96)
	at com.google.common.collect.Lists$TransformingRandomAccessList.get(Lists.java:638)
	at sklearn.pipeline.PipelineUtil.getHead(PipelineUtil.java:36)
	at sklearn2pmml.pipeline.PMMLPipeline.encodePMML(PMMLPipeline.java:138)
	at com.sklearn2pmml.Main.run(Main.java:91)
	at com.sklearn2pmml.Main.main(Main.java:66)
Caused by: java.lang.ClassCastException: Cannot cast net.razorvine.pickle.objects.ClassDict to sklearn.Transformer
	at java.lang.Class.cast(Class.java:3369)
	at org.jpmml.python.CastFunction.apply(CastFunction.java:45)
	... 7 more
```

This result was obtained with `sklearn2pmml` ver `0.91.0` and `scikit-learn` ver `1.2.1`.

**Comment by kota7** [comment]
A minute of thought get me this idea: Probably the hierarchical clustering algorithm is not very relevant here because it is does not `fit` method?  In other words the algorithm computes the clustering of the given training data, but does not have a capability of assigning cluster labels to unseen data?

**Comment by vruusmann** [comment]
> Probably the hierarchical clustering algorithm is not very relevant here because it is does not fit method? 

The PMML representation is about capturing the "business logic" of `predict(X)` methods.

And as you point out, the `AgglomerativeClustering` model type is atypical, because it has a `fit(X)` method, but no `predict(X)` method (it has `fit_predict(X)`, but it seems to be no-op?).

So, marking this issue as "conceptually not supported" (ie. the underlying Scikit-Learn estimator is missing a `predict(X)` method).

If you want to merge clusters, then perhaps you can build some sort of predict-able model ensemble for that? The first step would be `KMeans`, and the next step would be some kind of merging model. Perhaps a tree model.
